File: python/elizaos_plugin_whatsapp/providers/chat_state.py
"""Chat state provider for WhatsApp."""

import logging
from datetime import datetime
from typing import Any

from elizaos.types import Memory, Provider, State

from elizaos_plugin_whatsapp.types import WhatsAppChatState

logger = logging.getLogger(__name__)

WHATSAPP_SERVICE_NAME = "whatsapp"


async def get_chat_state(
    runtime: Any,
    message: Memory,
    state: State | None = None,
) -> str:
    """Gets the chat state for WhatsApp."""
    service = runtime.get_service(WHATSAPP_SERVICE_NAME)

    if not service or not service.is_running:
        return ""

    try:
        room = await runtime.get_room(message.room_id)
        if not room or not room.channel_id:
            return ""

        if room.source != "whatsapp":
            return ""

        chat_state = service.get_chat_state(room.channel_id)
        if not chat_state:
            return ""

        return _format_chat_state(chat_state)

    except Exception as e:
        logger.debug("Failed to get WhatsApp chat state: %s", e)
        return ""


def _format_chat_state(state: WhatsAppChatState) -> str:
    """Formats the chat state for inclusion in prompts."""
    lines = [
        "# WhatsApp Chat Context",
        "",
        f"- Contact: {state.contact_wa_id}",
    ]

    if state.contact_name:
        lines.append(f"- Name: {state.contact_name}")

    if state.last_message_at:
        dt = datetime.fromtimestamp(state.last_message_at)
        lines.append(f"- Last Message: {dt.strftime('%Y-%m-%d %H:%M:%S')}")

    lines.append("")
    lines.append("Note: This conversation is on WhatsApp. Be helpful and concise.")

    return "\n".join(lines)


chat_state_provider = Provider(
    name="WHATSAPP_CHAT_STATE",
    description="Provides information about the current WhatsApp chat context",
    get=get_chat_state,
)

File: rust/src/providers/chat_state.rs
//! Chat state provider for WhatsApp

use crate::service::WhatsAppService;
use crate::types::WhatsAppChatState;
use std::sync::Arc;

/// Provider that supplies WhatsApp chat context
pub struct ChatStateProvider {
    service: Option<Arc<WhatsAppService>>,
}

impl ChatStateProvider {
    pub fn new() -> Self {
        Self { service: None }
    }

    /// Create with a service reference
    pub fn with_service(service: Arc<WhatsAppService>) -> Self {
        Self {
            service: Some(service),
        }
    }

    /// Get the provider name
    pub fn name(&self) -> &str {
        "WHATSAPP_CHAT_STATE"
    }

    /// Get the provider description
    pub fn description(&self) -> &str {
        "Provides information about the current WhatsApp chat context"
    }

    /// Get chat state for a given contact
    pub async fn get_chat_state(&self, contact_id: &str) -> Option<String> {
        let service = self.service.as_ref()?;
        let chat_state = service.get_chat_state(contact_id).await?;
        Some(format_chat_state(&chat_state))
    }
}

impl Default for ChatStateProvider {
    fn default() -> Self {
        Self::new()
    }
}

fn format_chat_state(state: &WhatsAppChatState) -> String {
    let mut lines = vec![
        "# WhatsApp Chat Context".to_string(),
        String::new(),
        format!("- Contact: {}", state.contact_wa_id),
    ];

    if let Some(ref name) = state.contact_name {
        lines.push(format!("- Name: {}", name));
    }

    if let Some(timestamp) = state.last_message_at {
        lines.push(format!("- Last Message: {} (timestamp)", timestamp));
    }

    lines.push(String::new());
    lines.push("Note: This conversation is on WhatsApp. Be helpful and concise.".to_string());

    lines.join("\n")
}
